File: js/src/mcp/tools/browserTools.ts
import { chromium, type Browser, type BrowserContext, type Frame, type Page, type Route } from 'playwright';

import {
  createBrowserSecurityPolicy,
  type BrowserSecurityPolicy,
  resolveScreenshotPath,
  validateNavigationUrl
} from '../security.js';

export interface BrowserActionResult {
  success: boolean;
  message: string;
  url?: string;
  value?: string;
  screenshotPath?: string;
}

export interface NavigateParams { url: string; }
export interface ClickParams { selector: string; }
export interface FillParams { selector: string; value: string; }
export interface GetTextParams { selector: string; }
export interface ScreenshotParams { path?: string; }

export interface BrowserTools {
  navigate(params: NavigateParams): Promise<BrowserActionResult>;
  click(params: ClickParams): Promise<BrowserActionResult>;
  fill(params: FillParams): Promise<BrowserActionResult>;
  getText(params: GetTextParams): Promise<BrowserActionResult>;
  screenshot(params: ScreenshotParams): Promise<BrowserActionResult>;
}

export class BrowserController {
  private browser?: Browser;
  private context?: BrowserContext;
  private page?: Page;
  private pagePromise?: Promise<Page>;
  private bootstrapPage?: Page;
  private documentViolation?: Error;
  readonly policy: BrowserSecurityPolicy;

  constructor(policy: BrowserSecurityPolicy = createBrowserSecurityPolicy()) {
    this.policy = policy;
  }

  async getPage(): Promise<Page> {
    this.assertDocumentNavigationsAllowed();
    if (this.page) return this.page;
    this.pagePromise ??= this.createPage();

    try {
      this.page = await this.pagePromise;
      this.assertDocumentNavigationsAllowed();
      return this.page;
    } finally {
      this.pagePromise = undefined;
    }
  }

  validateNavigation(url: string): string {
    return validateNavigationUrl(url, this.policy.allowedOrigins);
  }

  resolveScreenshotPath(filename?: string): string {
    return resolveScreenshotPath(this.policy, filename);
  }

  assertDocumentNavigationsAllowed(): void {
    if (this.documentViolation) {
      throw this.documentViolation;
    }
  }

  private async createPage(): Promise<Page> {
    this.browser = await chromium.launch({ headless: true });
    this.context = await this.browser.newContext({
      acceptDownloads: false,
      baseURL: this.policy.baseURL,
      serviceWorkers: 'block'
    });
    await this.context.route('**/*', (route) => this.guardNavigation(route));
    const page = await this.context.newPage();
    this.bootstrapPage = page;
    this.context.on('framenavigated', (frame) => this.guardDocumentNavigation(frame));
    return page;
  }

  private async guardNavigation(route: Route): Promise<void> {
    const request = route.request();
    if (request.isNavigationRequest()) {
      try {
        validateNavigationUrl(request.url(), this.policy.allowedOrigins);
      } catch {
        await route.abort('blockedbyclient');
        return;
      }
    }
    await route.continue();
  }

  private guardDocumentNavigation(frame: Frame): void {
    if (frame.page() === this.bootstrapPage && frame === this.bootstrapPage.mainFrame()) {
      if (frame.url() === 'about:blank') return;
      this.bootstrapPage = undefined;
    }

    try {
      validateNavigationUrl(frame.url(), this.policy.allowedOrigins);
    } catch {
      this.documentViolation ??= new Error(`Document navigation blocked: ${frame.url()} is not allowed`);
      void frame.page().close().catch(() => undefined);
    }
  }

  async dispose(): Promise<void> {
    await this.context?.close();
    await this.browser?.close();
    this.context = undefined;
    this.page = undefined;
    this.pagePromise = undefined;
    this.bootstrapPage = undefined;
    this.browser = undefined;
    this.documentViolation = undefined;
  }
}

const requireNonEmpty = (value: string, label: string): void => {
  if (!value.trim()) {
    throw new Error(`${label} must not be empty`);
  }
};

export const createBrowserTools = (controller: BrowserController): BrowserTools => {
  return {
    navigate: async ({ url }: NavigateParams): Promise<BrowserActionResult> => {
      const validatedUrl = controller.validateNavigation(url);
      const page = await controller.getPage();
      await page.goto(validatedUrl);
      controller.assertDocumentNavigationsAllowed();
      return { success: true, message: 'Navigated', url: page.url() };
    },
    click: async ({ selector }: ClickParams): Promise<BrowserActionResult> => {
      requireNonEmpty(selector, 'Selector');
      const page = await controller.getPage();
      await page.click(selector);
      controller.assertDocumentNavigationsAllowed();
      return { success: true, message: `Clicked ${selector}`, url: page.url() };
    },
    fill: async ({ selector, value }: FillParams): Promise<BrowserActionResult> => {
      requireNonEmpty(selector, 'Selector');
      const page = await controller.getPage();
      await page.fill(selector, value);
      controller.assertDocumentNavigationsAllowed();
      return { success: true, message: `Filled ${selector}` };
    },
    getText: async ({ selector }: GetTextParams): Promise<BrowserActionResult> => {
      requireNonEmpty(selector, 'Selector');
      const page = await controller.getPage();
      const content = await page.textContent(selector);
      controller.assertDocumentNavigationsAllowed();
      return { success: true, message: 'Text retrieved', value: content ?? '' };
    },
    screenshot: async ({ path: screenshotPath }: ScreenshotParams): Promise<BrowserActionResult> => {
      const resolvedPath = controller.resolveScreenshotPath(screenshotPath);
      const page = await controller.getPage();
      await page.screenshot({ path: resolvedPath, fullPage: true });
      controller.assertDocumentNavigationsAllowed();
      return { success: true, message: 'Screenshot captured', screenshotPath: resolvedPath };
    }
  };
};

File: py/src/playwright_adventures/mcp_server/tools/browser_tools.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass

from playwright.async_api import Browser, BrowserContext, Page, Playwright, Route, async_playwright

from ..security import BrowserSecurityPolicy


@dataclass
class BrowserResult:
    success: bool
    message: str
    url: str | None = None
    value: str | None = None
    screenshot_path: str | None = None


class BrowserSession:
    def __init__(self, policy: BrowserSecurityPolicy | None = None) -> None:
        self._browser: Browser | None = None
        self._context: BrowserContext | None = None
        self._page: Page | None = None
        self._playwright: Playwright | None = None
        self._lock = asyncio.Lock()
        self.policy = policy or BrowserSecurityPolicy.from_environment()

    async def get_page(self) -> Page:
        async with self._lock:
            if self._page:
                return self._page

            playwright = await async_playwright().start()
            self._playwright = playwright
            self._browser = await playwright.chromium.launch(headless=True)
            self._context = await self._browser.new_context(
                accept_downloads=False,
                base_url=self.policy.base_url,
                service_workers="block",
            )
            await self._context.route("**/*", self._guard_navigation)
            self._page = await self._context.new_page()
            return self._page

    async def _guard_navigation(self, route: Route) -> None:
        request = route.request
        if request.is_navigation_request():
            try:
                self.policy.validate_navigation_url(request.url)
            except ValueError:
                await route.abort(error_code="blockedbyclient")
                return
        await route.continue_()

    async def close(self) -> None:
        async with self._lock:
            if self._context:
                await self._context.close()
            if self._browser:
                await self._browser.close()
            if self._playwright:
                await self._playwright.stop()
            self._context = None
            self._browser = None
            self._page = None
            self._playwright = None


def _require_non_empty(value: str, label: str) -> None:
    if not value.strip():
        raise ValueError(f"{label} must not be empty")


class BrowserTools:
    def __init__(self, session: BrowserSession) -> None:
        self.session = session

    async def browser_navigate(self, url: str) -> BrowserResult:
        validated_url = self.session.policy.validate_navigation_url(url)
        page = await self.session.get_page()
        await page.goto(validated_url)
        return BrowserResult(success=True, message="Navigated", url=page.url)

    async def browser_click(self, selector: str) -> BrowserResult:
        _require_non_empty(selector, "Selector")
        page = await self.session.get_page()
        await page.click(selector)
        return BrowserResult(success=True, message=f"Clicked {selector}", url=page.url)

    async def browser_fill(self, selector: str, value: str) -> BrowserResult:
        _require_non_empty(selector, "Selector")
        page = await self.session.get_page()
        await page.fill(selector, value)
        return BrowserResult(success=True, message=f"Filled {selector}")

    async def browser_get_text(self, selector: str) -> BrowserResult:
        _require_non_empty(selector, "Selector")
        page = await self.session.get_page()
        content = await page.text_content(selector)
        return BrowserResult(success=True, message="Text retrieved", value=content or "")

    async def browser_screenshot(self, path: str | None = None) -> BrowserResult:
        loop = asyncio.get_running_loop()
        resolved_path = self.session.policy.resolve_screenshot_path(path, int(loop.time() * 1000))
        page = await self.session.get_page()
        await page.screenshot(path=str(resolved_path), full_page=True)
        return BrowserResult(success=True, message="Screenshot captured", screenshot_path=str(resolved_path))
